designs/aws/custom-user-data-and-amis.md

```diff
@@ -0,0 +1,285 @@
+cluster-certificate = 'LS0tLS....'
+cluster-name = 'my-cluster'
+settings.kubernetes.api-server = 'apiServerEndpoint'
+
+[settings.kubernetes.eviction-hard]
+"memory.available" = "15%"
+
+[settings.kubernetes.node-labels]
+'karpenter.sh/capacity-type' = 'on-demand'
+'karpenter.sh/provisioner-name' = 'default'
+'my-custom-label' = 'santaClaus'
+```
+
+From the output of the merged output, you can see that Karpenter will override any fields that it considers necessary (`node-labels` based on what needs to be present as per the incoming pending pod, `cluster-certificate` etc). All other fields will be copied over as is from the user's TOML content.
+
+* You can address `CASE2` via this mode. `CASE1` isn’t very applicable to BottleRocket nodes.
+* ManagedNodegroups performs a similar merge so there is precedence and is feasible.
+    * The merge is necessary because TOML doesn’t support [duplicate keys](https://github.com/toml-lang/toml/issues/697) as part of its spec, so you can’t just concatenate two TOML files.
+* To ensure that we don’t perform incorrect bin-packing, Karpenter might need to parse the UserData and extract `settings.kubernetes.system-reserved` and `settings.kubernetes.kube-reserved` which is needed to better estimate nodeAllocatable. This might mean Karpenter maintains state / a cache per Karpenter provisioner.
+    * Alternatively, we can introduce a new field to our provisioner that helps hint how to binpack. This is further discussed in the following section since this limitation applies to the EKS AL2 AMIs too.
+
+
+**For EKS-optimized AL2 AMIFamily**
+
+AL2 / Linux AMIs is more popularly used. UserData is executed by [cloud-init](https://cloudinit.readthedocs.io/en/latest/) and the user data is expected to be shell scripts / cloud-init directives. Merging shell scripts can be messy, so the default pattern used by other AWS services is to leverage MIME multi-part data.
+
+There are different options we have here -
+
+*Option 1 - MIME parts are run in the order -  `UserManaged` → `Karpenter Managed`*
+
+```
+MIME-Version: 1.0
+Content-Type: multipart/mixed; boundary="==MYBOUNDARY=="
+
+--==MYBOUNDARY==
+Content-Type: text/x-shellscript; charset="us-ascii"
+
+#!/bin/bash
+echo "Running custom user data script that was in the ProvisionerSpec"
+
+--==MYBOUNDARY==
+Content-Type: text/x-shellscript; charset="us-ascii"
+
+#!/bin/bash
+set -ex
+B64_CLUSTER_CA=base64data
+API_SERVER_URL=apiServerEndpoint
+K8S_CLUSTER_DNS_IP=k8sClusterDnsIp
+/etc/eks/bootstrap.sh clusterName
+  --kubelet-extra-args '--node-labels=karpenter.sh/provisioner-name=default,karpenter.sh/capacity-type=on-demand' \
+  --b64-cluster-ca $B64_CLUSTER_CA \
+  --apiserver-endpoint $API_SERVER_URL \
+  --dns-cluster-ip $K8S_CLUSTER_DNS_IP
+
+--==MYBOUNDARY==--
+```
+
+* Users can address CASE1 very well through this mode. They can leave Karpenter to figure how to bootstrap the worker node and install whatever they need on the worker node.
```

Comment:
I think this can be improved to "leave it to Karpenter" or simply "let Karpenter"